src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,335 @@
+        if !arc_self.supressed.load(Ordering::SeqCst) {
+            arc_self.activator.activate().unwrap();
+        }
+    }
+}
+
+/// Async handle to an operator's input stream
+pub struct AsyncInputHandle<T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> {
+    /// The underying synchronous input handle
+    sync_handle: InputHandleCore<T, D, P>,
+    /// Frontier information of input streams shared with the operator. Each frontier is paired
+    /// with a flag indicating whether or not the handle has seen the updated frontier.
+    shared_frontiers: Rc<RefCell<Vec<(Antichain<T>, bool)>>>,
+    /// The index of this handle into the shared frontiers vector
+    index: usize,
+    /// Reference to the reactor queue of this input handle where Wakers can be registered
+    reactor_registry: Rc<RefCell<Vec<Waker>>>,
+}
+
+impl<T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> AsyncInputHandle<T, D, P> {
+    /// Produces a future that will resolve to the next event of this input stream
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel-safe
+    pub fn next(&mut self) -> impl Future<Output = Option<Event<'_, T, D>>> + '_ {
+        NextFut { handle: Some(self) }
+    }
+}
+
+/// The future returned by `AsyncInputHandle::next`
+struct NextFut<'a, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> {
+    handle: Option<&'a mut AsyncInputHandle<T, D, P>>,
+}
+
+/// An event of an input stream
+pub enum Event<'a, T: Timestamp, D> {
+    /// A data event
+    Data(Capability<T>, RefOrMut<'a, D>),
+    /// A progress event
+    Progress(Antichain<T>),
+}
+
+impl<'a, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future for NextFut<'a, T, D, P> {
+    type Output = Option<Event<'a, T, D>>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let handle = self.handle.take().expect("future polled after completion");
+        match handle.sync_handle.next() {
+            Some((cap, data)) => {
+                // We must retain the capability because timely isn't prepared to handle
+                // CapabilityRefs that are retained between operator invocations. This situation is
+                // very difficult to happen in safe sync Rust but trivial with an async operator so
+                // it's best to prevent the footgun.
+                //
+                // However this introduces a new footgun. `Capability::retain()` retains the
+                // capability only for the first output, meaning that an async operator can still
+                // panic if the retained capability survives an await point and then is later used
+                // to send messages to some other output of the operator. Usually operators have a
+                // single output so this seems like less serious footgun than the other one.
+                //
+                // The above will be solved once <link to PR> is merged where the CapabilityRef
```

Comment:
This refers to this PR https://github.com/TimelyDataflow/timely-dataflow/pull/429